lib/spack/spack/provider_index.py

```diff
@@ -2,54 +2,134 @@
+    for lspec, rspec in itertools.product(lmap, rmap):
+        try:
+            constrained = lspec.constrained(rspec)
+        except spack.error.UnsatisfiableSpecError:
+            continue
+
+        # lp and rp are left and right provider specs.
+        for lp_spec, rp_spec in itertools.product(lmap[lspec], rmap[rspec]):
+            if lp_spec.name == rp_spec.name:
+                try:
+                    const = lp_spec.constrained(rp_spec, deps=False)
+                    result.setdefault(constrained, set()).add(const)
+                except spack.error.UnsatisfiableSpecError:
+                    continue
+    return result
+
+
+class _IndexBase(object):
+    #: This is a dict of dicts used for finding providers of particular
+    #: virtual dependencies. The dict of dicts looks like:
+    #:
+    #:    { vpkg name :
+    #:        { full vpkg spec : set(packages providing spec) } }
+    #:
+    #: Callers can use this to first find which packages provide a vpkg,
+    #: then find a matching full spec.  e.g., in this scenario:
+    #:
+    #:    { 'mpi' :
+    #:        { mpi@:1.1 : set([mpich]),
+    #:          mpi@:2.3 : set([mpich2@1.9:]) } }
+    #:
+    #: Calling providers_for(spec) will find specs that provide a
+    #: matching implementation of MPI. Derived class need to construct
+    #: this attribute according to the semantics above.
+    providers = None
+
+    def providers_for(self, virtual_spec):
+        """Return a list of specs of all packages that provide virtual
+        packages with the supplied spec.
+
+        Args:
+            virtual_spec: virtual spec to be provided
+        """
+        result = set()
+        # Allow string names to be passed as input, as well as specs
+        if type(virtual_spec) == str:
+            virtual_spec = spack.spec.Spec(virtual_spec)
 
-       Callers can use this to first find which packages provide a vpkg,
-       then find a matching full spec.  e.g., in this scenario:
+        # Add all the providers that satisfy the vpkg spec.
+        if virtual_spec.name in self.providers:
+            for p_spec, spec_set in self.providers[virtual_spec.name].items():
+                if p_spec.satisfies(virtual_spec, deps=False):
+                    result.update(spec_set)
 
-       { 'mpi' :
-           { mpi@:1.1 : set([mpich]),
-             mpi@:2.3 : set([mpich2@1.9:]) } }
+        # Return providers in order. Defensively copy.
+        return sorted(s.copy() for s in result)
 
-       Calling providers_for(spec) will find specs that provide a
-       matching implementation of MPI.
+    def __contains__(self, name):
+        return name in self.providers
 
-    """
+    def satisfies(self, other):
+        """Return True if the providers of virtual specs are compatible,
+        False otherwise.
 
-    def __init__(self, specs=None, restrict=False):
-        """Create a new ProviderIndex.
+        Args:
+            other: another provider index
+        """
```

Comment:
Would it be more appropriate for the description to be something like: 

``Determine if the providers of virtual specs are compatible``

and add the standard `Returns:` description after the `Args`?